geth/api/backend.go

```diff
@@ -77,120 +80,146 @@ func (m *StatusBackend) IsNodeRunning() bool {
 	}
-
 	if err := m.accountManager.ReSelectAccount(); err != nil {
 		log.Error("Reselect account failed", "err", err)
 	}
 	log.Info("Account reselected")
-
-	close(backendReady)
 	signal.Send(signal.Envelope{
 		Type:  signal.EventNodeReady,
 		Event: struct{}{},
 	})
 }
 
 // StopNode stop Status node. Stopped node cannot be resumed.
-func (m *StatusBackend) StopNode() (<-chan struct{}, error) {
-	m.Lock()
-	defer m.Unlock()
+func (m *StatusBackend) StopNode() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return m.stopNode()
+}
 
-	if m.nodeReady == nil {
-		return nil, node.ErrNoRunningNode
+func (m *StatusBackend) stopNode() error {
+	if !m.IsNodeRunning() {
+		return node.ErrNoRunningNode
 	}
-	<-m.nodeReady
-
 	m.txQueueManager.Stop()
 	m.jailManager.Stop()
-
-	nodeStopped, err := m.nodeManager.StopNode()
+	err := m.nodeManager.StopNode()
 	if err != nil {
-		return nil, err
+		return err
 	}
-
-	backendStopped := make(chan struct{}, 1)
-	go func() {
-		<-nodeStopped
-		m.Lock()
-		m.nodeReady = nil
-		m.Unlock()
-		close(backendStopped)
-	}()
-
-	return backendStopped, nil
+	signal.Send(signal.Envelope{
+		Type:  signal.EventNodeStopped,
+		Event: struct{}{},
+	})
+	return nil
 }
 
 // RestartNode restart running Status node, fails if node is not running
-func (m *StatusBackend) RestartNode() (<-chan struct{}, error) {
-	m.Lock()
-	defer m.Unlock()
-
-	if m.nodeReady == nil {
-		return nil, node.ErrNoRunningNode
+func (m *StatusBackend) RestartNode() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if !m.IsNodeRunning() {
+		return node.ErrNoRunningNode
 	}
-	<-m.nodeReady
-
-	nodeRestarted, err := m.nodeManager.RestartNode()
+	config, err := m.nodeManager.NodeConfig()
 	if err != nil {
-		return nil, err
+		return err
 	}
-
-	m.nodeReady = make(chan struct{}, 1)
-	go m.onNodeStart(nodeRestarted, m.nodeReady) // waits on nodeRestarted, writes to backendReady
-
-	return m.nodeReady, err
+	newcfg := *config
+	if err := m.stopNode(); err != nil {
+		return err
+	}
+	if err := m.startNode(&newcfg); err != nil {
+		return err
+	}
+	return err
 }
 
 // ResetChainData remove chain data from data directory.
 // Node is stopped, and new node is started, with clean data directory.
-func (m *StatusBackend) ResetChainData() (<-chan struct{}, error) {
-	m.Lock()
-	defer m.Unlock()
-
-	if m.nodeReady == nil {
-		return nil, node.ErrNoRunningNode
-	}
-	<-m.nodeReady
-
-	nodeReset, err := m.nodeManager.ResetChainData()
+func (m *StatusBackend) ResetChainData() error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	config, err := m.nodeManager.NodeConfig()
 	if err != nil {
-		return nil, err
+		return err
 	}
-
-	m.nodeReady = make(chan struct{}, 1)
-	go m.onNodeStart(nodeReset, m.nodeReady) // waits on nodeReset, writes to backendReady
-
-	return m.nodeReady, err
+	newcfg := *config
+	if err := m.stopNode(); err != nil {
+		return err
+	}
+	chainDataDir := filepath.Join(config.DataDir, config.Name, "lightchaindata")
```

Comment:
Hm, not sure why we remove chain data dir in the backend now instead of `NodeManager`.

Reply:
i moved restart node and reset chain data to simplify node manager interface